### PR TITLE
bpo-40334: Improve various PEG-Parser related stuff

### PR DESCRIPTION
**Pull request by lysnikolaou**
These are all related to @vstinner's review of the PEP 617 PR.
The most important changes include moving `pegen_interface.h`
into `Include/internal` and refactoring some code in
`Parser/pegen/pegen.c`.

https://bugs.python.org/issue40334

### CODE REVIEW

**Comment by vstinner** [comment]
Oh, this PR partially conflicts with my PR #19670. Would you mind to review it?

**Comment by lysnikolaou** [comment]
> Oh, this PR partially conflicts with my PR #19670. Would you mind to review it?

Of course. Let me just first remove the conflicts from this one.

File: Parser/pegen/pegen.c [hunk 1]
```diff
@@ -530,13 +540,21 @@ _PyPegen_fill_token(Parser *p)
 
     if (p->fill == p->size) {
         int newsize = p->size * 2;
-        p->tokens = PyMem_Realloc(p->tokens, newsize * sizeof(Token *));
-        if (p->tokens == NULL) {
-            PyErr_Format(PyExc_MemoryError, "Realloc tokens failed");
+        Token **new_tokens = PyMem_Realloc(p->tokens, newsize * sizeof(Token *));
+        if (new_tokens == NULL) {
+            PyMem_Free(p->tokens);
+            p->tokens = NULL;
```
Comment (vstinner) [inline Parser/pegen/pegen.c]: Maybe the caller already frees token when it cleans up the parser object? I didn't check how a parser is destroyed.

File: Parser/pegen/pegen.c [hunk 2]
```diff
@@ -530,13 +540,21 @@ _PyPegen_fill_token(Parser *p)
...
             return -1;
         }
+        else {
+            p->tokens = new_tokens;
+        }
         for (int i = p->size; i < newsize; i++) {
             p->tokens[i] = PyMem_Malloc(sizeof(Token));
+            if (p->tokens[i] == NULL) {
```
Comment (vstinner) [inline Parser/pegen/pegen.c:560]: You need to update  p->size here. Maybe set it it i or i-1, I'm not sure :-p

File: Include/internal/pegen_interface.h [hunk 3]
```diff
@@ -1,4 +1,3 @@
-#ifndef Py_LIMITED_API
 #ifndef Py_PEGENINTERFACE
 #define Py_PEGENINTERFACE
```
Comment (vstinner) [inline Include/internal/pegen_interface.h:2]: Oh, it's even better if the header file can be marked as internal!

I suggest to add the following guard (see for example pycore_pymem.h):

```
#ifndef Py_BUILD_CORE
#  error "this header requires Py_BUILD_CORE define"
#endif
```

File: Parser/pegen/pegen.c (hunk 2 above)
Comment (lysnikolaou) [reply Parser/pegen/pegen.c:560]: Is that needed if we fail hard here? If `fill_token` returns -1, the parser fails for every rule until it backtracks out the start function and it destroys the `Parser` struct.

Comment (vstinner) [reply Parser/pegen/pegen.c:560]: When the parser is destroyed, you have to free memory of tokens between `p->size` and `<i value when the loop stopped>`, not only `0` to `p->size`.

Comment (lysnikolaou) [reply Parser/pegen/pegen.c:560]: Of course!

File: Parser/pegen/pegen.c [hunk 4]
```diff
@@ -337,9 +352,6 @@ tokenizer_error(Parser *p)
             errtype = PyExc_IndentationError;
             msg = "too many levels of indentation";
             break;
-        case E_DECODE:
-            raise_decode_error(p);
-            return -1;
```
Comment (vstinner) [inline Parser/pegen/pegen.c:342]: Alright, I didn't understand this code path. I now understand better raise_decode_error() with the assertion, thanks ;-)

File: Parser/pegen/pegen.c [hunk 5]
```diff
@@ -530,13 +547,21 @@ _PyPegen_fill_token(Parser *p)
...
+            PyErr_NoMemory();
             return -1;
         }
+        else {
+            p->tokens = new_tokens;
+        }
         for (int i = p->size; i < newsize; i++) {
             p->tokens[i] = PyMem_Malloc(sizeof(Token));
+            if (p->tokens[i] == NULL) {
+                p->size = i - 1; // Needed, in order to cleanup correctly after parser fails
```
Comment (vstinner) [inline Parser/pegen/pegen.c]: Are you use about `i -1`? _PyPegen_Parser_Free():

```
    for (int i = 0; i < p->size; i++) {
        PyMem_Free(p->tokens[i]);
    }
```

For example, if the loop fails at the first iteration, you replace p->size=SIZE with p->size=(SIZE - 1), no?

File: Parser/pegen/pegen.c [hunk 6]
```diff
@@ -878,8 +897,7 @@ _PyPegen_Parser_New(struct tok_state *tok, int start_rule, int *errcode, PyArena
 {
     Parser *p = PyMem_Malloc(sizeof(Parser));
     if (p == NULL) {
-        PyErr_Format(PyExc_MemoryError, "Out of memory for Parser");
-        return NULL;
+        return (void *) PyErr_NoMemory();
```
Comment (vstinner) [inline Parser/pegen/pegen.c]: If you cast explicitly, I would prefer to cast to `Parser*`. Same coment for other error paths below.

File: Parser/pegen/pegen.c [hunk 7]
```diff
@@ -1177,7 +1199,7 @@ _PyPegen_seq_count_dots(asdl_seq *seq)
                 number_of_dots += 1;
                 break;
             default:
-                assert(current_expr->type == ELLIPSIS || current_expr->type == DOT);
+                Py_UNREACHABLE();
```
Comment (vstinner) [inline Parser/pegen/pegen.c:1202]: Ah thanks :-)

File: Parser/pegen/pegen.c (hunk 5 above)
Comment (lysnikolaou) [reply Parser/pegen/pegen.c]: You're very right. I was thinking that the loop in `_PyPegen_Parser_Free` was going up to and including `p->size` for some reason. Pushing a fix.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by lysnikolaou** [comment]
Thanks for the help, @vstinner!

**Pull request merged**